Build fails with Ubuntu 24.04 / gcc 13.2.0

Trying to update GDAL Docker image to Ubuntu 24.04, I get the following compilation error:
```
Cloning into 'QB3'...
-- The CXX compiler identification is GNU 13.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: /QB3/QB3lib/build
[ 66%] Building CXX object CMakeFiles/libQB3.dir/QB3encode.cpp.o
[ 66%] Building CXX object CMakeFiles/libQB3.dir/QB3decode.cpp.o
In file included from /usr/include/c++/13/vector:63,
                 from /QB3/QB3lib/QB3decode.cpp:22:
/usr/include/c++/13/bits/allocator.h: In destructor 'std::_Vector_base<unsigned char, std::allocator<unsigned char> >::_Vector_impl::~_Vector_impl()':
/usr/include/c++/13/bits/allocator.h:184:7: error: inlining failed in call to 'always_inline' 'std::allocator< <template-parameter-1-1> >::~allocator() noexcept [with _Tp = unsigned char]': target specific option mismatch
  184 |       ~allocator() _GLIBCXX_NOTHROW { }
      |       ^
In file included from /usr/include/c++/13/vector:66:
/usr/include/c++/13/bits/stl_vector.h:133:14: note: called from here
  133 |       struct _Vector_impl
      |              ^~~~~~~~~~~~
make[2]: *** [CMakeFiles/libQB3.dir/build.make:90: CMakeFiles/libQB3.dir/QB3decode.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /usr/include/c++/13/vector:63,
                 from /QB3/QB3lib/QB3encode.cpp:21:
/usr/include/c++/13/bits/allocator.h: In destructor 'std::_Vector_base<unsigned char, std::allocator<unsigned char> >::_Vector_impl::~_Vector_impl()':
/usr/include/c++/13/bits/allocator.h:184:7: error: inlining failed in call to 'always_inline' 'std::allocator< <template-parameter-1-1> >::~allocator() noexcept [with _Tp = unsigned char]': target specific option mismatch
  184 |       ~allocator() _GLIBCXX_NOTHROW { }
      |       ^
In file included from /usr/include/c++/13/vector:66:
/usr/include/c++/13/bits/stl_vector.h:133:14: note: called from here
  133 |       struct _Vector_impl
      |              ^~~~~~~~~~~~
make[2]: *** [CMakeFiles/libQB3.dir/build.make:76: CMakeFiles/libQB3.dir/QB3encode.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/libQB3.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

how to reproduce:
inside ``docker run --rm -it ubuntu:24.04``:
```
apt update
apt install git cmake g++
git clone https://github.com/lucianpls/QB3.git
cd QB3/QB3lib
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release .. 
make -j$(nproc)
```

--- a/QB3lib/CMakeLists.txt
+++ b/QB3lib/CMakeLists.txt
@@ -20,9 +20,22 @@ option(BUILD_SHARED_LIBS "Library style" ON)
 set(namespace "QB3")
 add_library(${PROJECT_NAME})
 
-# On AMD64 uncomment to use avx2 -> faster code
-# target_compile_options(${PROJECT_NAME} PRIVATE $<$<CXX_COMPILER_ID:MSVC>:/arch:AVX2>)
-# target_compile_options(${PROJECT_NAME} PRIVATE $<$<CXX_COMPILER_ID:GNU>:-mavx2>)
+if("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "x86_64" AND ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU"))
+    option(ENABLE_SSE4 "Enable SSE4 optimizations" ON)
+    if (ENABLE_SSE4)
+        target_compile_options(${PROJECT_NAME} PRIVATE "-msse4")
+    endif()
+
+    option(ENABLE_AVX2 "Enable AVX2 optimizations" OFF)
+    if (ENABLE_AVX2)
+        target_compile_options(${PROJECT_NAME} PRIVATE "-mavx2")
+    endif()
+elseif("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "AMD64" AND MSVC)
+    option(ENABLE_AVX2 "Enable AVX2 optimizations" OFF)
+    if (ENABLE_AVX2)
+        target_compile_options(${PROJECT_NAME} PRIVATE "/arch:AVX2")
+    endif()
+endif()
 
 target_sources(${PROJECT_NAME} 
     PRIVATE QB3encode.cpp QB3encode.h QB3decode.cpp QB3decode.h QB3common.h bitstream.h QB3.h

--- a/QB3lib/QB3common.h
+++ b/QB3lib/QB3common.h
@@ -16,6 +16,7 @@ Contributors:  Lucian Plesea
 */
 
 #pragma once
+
 #include "QB3.h"
 #include "bitstream.h"
 #include <cinttypes>
@@ -32,11 +33,6 @@ constexpr auto TBLMASK(0xfffull);
 // Block is 4x4 pixels
 constexpr size_t B(4);
 constexpr size_t B2(B * B);
-
-#if defined(__GNUC__) && defined(__x86_64__)
-// Comment out if binary is to run on processors without sse4
-#pragma GCC target("sse4")
-#endif
 
 #if defined(_WIN32)
 // blog2 of val, result is undefined for val == 0

